### PR TITLE
CircleCI config added :construction_worker:

### PR DESCRIPTION
**Pull request by NexFlare**
**Motivation**

To migrate from travis to circleci for better CI on github repos.

**Changes**

- Updated InfluxDB installation step in Dockerfile
- Added CircleCI config
- Reduced the sleep time in docker_compose script by adding while loop to check if nuoadmin container is up
- Added host variable in trace_test and stats_test
- Bumped up the nuoversion and updated the install steps in install_environment script

### CODE REVIEW

File: tests/ci/docker_compose.sh [hunk 1]
```diff
@@ -7,6 +7,13 @@ docker pull $DOCKER_IMAGE
...
 
-docker exec "$(docker container ls -f name=nuoadmin -q)" nuocmd check database --db-name hockey --check-running --num-processes 2 --timeout 300
+while [ "$(docker container inspect -f '{{.State.Running}}' "$(docker ps -f name=nuoadmin -q)")" = "false" ]
+do
+    sleep 1
+    echo "Sleep for a second"
+done
+
+
+docker exec "$(docker container ls -f name=nuoadmin -q)" nuocmd check database --db-name hockey --check-running --num-processes 2 --timeout 300
```
Comment (sivanov-nuodb) [inline tests/ci/docker_compose.sh:19]: Nit: You use directly `docker exec <container name>`.

File: tests/ci/install_nuocd.sh [hunk 2]
```diff
@@ -5,13 +5,14 @@ export NUOCMD_API_SERVER=http://localhost:8888
...
 sudo cp conf/nuodb.conf /etc/telegraf/telegraf.d
 sudo cp conf/outputs.conf /etc/telegraf/telegraf.d
 sudo cp bin/nuocd /usr/local/bin/nuocd
 
 sudo systemctl daemon-reload
 sudo systemctl stop telegraf
 
-/usr/bin/telegraf --config /etc/telegraf/telegraf.conf --config-directory /etc/telegraf/telegraf.d&
\ No newline at end of file
+/usr/bin/telegraf --config /etc/telegraf/telegraf.conf --config-directory /etc/telegraf/telegraf.d
```
Comment (sivanov-nuodb) [inline tests/ci/install_nuocd.sh:18]: I think that this command blocks and that is why it was run in the background.
Have you tried to run the tests in bare-metal?

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    steps:
+      - nuodb_collector_command
+    environment:
+      TEST_SUITE: "Bare Metal"
+
+workflows:
+  build-project:
+    jobs:
+      - build_n_run_docker
+      # - build_n_run_bare_metal
```
Comment (sivanov-nuodb) [inline .circleci/config.yml:43]: I think that we want to run the tests in both modes so you may want to uncomment this job which will be run in parallel.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+version: 2.1
+
+commands:
+  nuodb_collector_command:
```
Comment (sivanov-nuodb) [inline .circleci/config.yml:4]: What is the advantage of having a separate `command` vs having parametrized jobs?
For example, you can take a look at [go-test](https://github.com/nuodb/nuodb-helm-charts/blob/master/.circleci/config.yml) job.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      - run:
+          name : Take a nap
+          command: |
+            sleep 15
+      - run:
+          name : Configure and run tests
+          command: |
+            pip install -r test_requirements.txt
+            python ./run_tests.py
+            echo "Done"
```
Comment (sivanov-nuodb) [inline .circleci/config.yml:19]: Can you attach the JUnit test reports to CircleCI using `store_test_results` step?

**Review by sivanov-nuodb — COMMENTED** [review]
@NexFlare Thanks for adding the CircleCI config for this project!
I left some comments regarding testing in both Docker and bare-metal modes.

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+jobs:
+  build_n_run_docker:
+    machine:
+      image: ubuntu-2204:2023.04.2
+    steps:
+      - nuodb_collector_command
+    environment:
+      TEST_SUITE: Docker
+    resource_class: medium
```
Comment (sivanov-nuodb) [inline .circleci/config.yml:29]: I saw that there are plenty of free resources in the CircleCI report. Does `small` work?

File: tests/ci/install_nuocd.sh (hunk 2 above)
Comment (NexFlare) [reply tests/ci/install_nuocd.sh:18]: The test in bare metal were failing and going in an infinite loop. For reference https://app.circleci.com/pipelines/github/nuodb/nuodb-collector/25/workflows/526a19c0-dd4a-4fe8-88f8-aab736aee108/jobs/33

File: .circleci/config.yml (hunk 4 above)
Comment (NexFlare) [reply .circleci/config.yml:4]: Added command just for reusability. The only difference between bare_metal and docker job is the environment variable. I set the variable in the job.

File: .circleci/config.yml (hunk 6 above)
Comment (NexFlare) [reply .circleci/config.yml:29]: Nopes. The ubuntu image's smallest resource is medium

Comment (sivanov-nuodb) [reply .circleci/config.yml:29]: Sounds good.

**Review by sivanov-nuodb — APPROVED** [review]

File: tests/ci/docker_compose.sh [hunk 7]
```diff
@@ -7,6 +7,13 @@ docker pull $DOCKER_IMAGE
 
 docker-compose up -d
 
-sleep 120
+#check if nuoadmin is up
 
-docker exec "$(docker container ls -f name=nuoadmin -q)" nuocmd check database --db-name hockey --check-running --num-processes 2 --timeout 300
+while [ "$(docker container inspect -f '{{.State.Running}}' "$(docker ps -f name=nuoadmin -q)")" = "false" ]
```
[the remark this answers is not in the document]
Comment (NexFlare) [reply tests/ci/docker_compose.sh:12]: Added the mentioned command

File: .circleci/config.yml (hunk 5 above)
Comment (NexFlare) [reply .circleci/config.yml]: Thanks for the suggestion. I have added the test results.

File: tests/ci/docker_compose.sh [hunk 8]
```diff
@@ -7,6 +7,14 @@ docker pull $DOCKER_IMAGE
...
```
Comment (adriansuarez) [inline tests/ci/docker_compose.sh:12]: I think you should first check that `docker ps -f name=nuoadmin -q` actually returns anything, and then check its state. It is possible that the container has not actually been started.

File: tests/ci/docker_compose.sh (hunk 1 above)
Comment (adriansuarez) [reply tests/ci/docker_compose.sh:20]: I think the issue is that Docker Compose comes up with longer names for the containers. The name here is actually the hostname that gets supplied to the container, not the container name.

File: tests/ci/install_environment.sh [hunk 9]
```diff
@@ -1,29 +1,31 @@
...
-export NUOVERSION=4.0.7.1
+export NUOVERSION=5.0.3.3
+export TAR_VERSION=nuodb-${NUOVERSION}.linux.x86_64
 export INFLUX_VERSION=1.4.3
-
-sudo sh -c 'echo madvise > /sys/kernel/mm/transparent_hugepage/enabled'
-sudo sh -c 'echo madvise > /sys/kernel/mm/transparent_hugepage/defrag'
-wget -q http://ce-downloads.nuohub.org/nuodb-ce_${NUOVERSION}_amd64.deb --output-document=/var/tmp/nuodb.deb
-sudo dpkg -i /var/tmp/nuodb.deb
+echo "http://ce-downlaods.nuohub.org/${TAR_VERSION}.tar.gz"
```
Comment (adriansuarez) [inline tests/ci/install_environment.sh:6]: You can drop this print statement. If you want to see what the script is doing, you can invoke `set -x` at the beginning of the script.

File: tests/ci/install_environment.sh [hunk 10]
```diff
@@ -1,29 +1,31 @@
...
+wget http://ce-downloads.nuohub.org/${TAR_VERSION}.tar.gz --output-document=/var/tmp/nuodb.tgz
+tar xzf /var/tmp/nuodb.tgz
+export NUODB_HOME=${PWD}/${TAR_VERSION}
 sleep 5
```
Comment (adriansuarez) [inline tests/ci/install_environment.sh:10]: What is the sleep for? We haven't actually started anything yet, have we?

File: tests/ci/install_environment.sh [hunk 11]
```diff
@@ -1,29 +1,31 @@
...
 
-export PATH=$PATH:/opt/nuodb/bin
-sudo sed -i -e "/ssl.:/s/true/false/" /etc/nuodb/nuoadmin.conf
-sudo service nuoadmin start
-nuocmd create archive --db-name hockey --server-id nuoadmin-0 --archive-path /var/opt/nuodb/demo-archives/hockey
+export PATH=$PATH:${NUODB_HOME}/bin
+export NUO_SET_TLS=disable
+$NUODB_HOME/etc/nuoadmin tls $NUO_SET_TLS
```
Comment (adriansuarez) [inline tests/ci/install_environment.sh:14]: Nit: I don't think you need to make `NUO_SET_TLS` a variable.

File: tests/ci/install_environment.sh [hunk 12]
```diff
@@ -1,29 +1,31 @@
...
+$NUODB_HOME/etc/nuoadmin start
```
Comment (adriansuarez) [inline tests/ci/install_environment.sh:15]: There is some waiting in the service wrapper script, but you should use the same `nuocmd check servers` command that you used in the Docker Compose test.

File: tests/ci/install_nuocd.sh (hunk 2 above)
Comment (adriansuarez) [reply tests/ci/install_nuocd.sh:18]: I'm confused about how this ever could have worked. The command puts the `telegraf` service in the background, but it does not invoke it with `nohup`, so I'm pretty sure it should will just be shutdown as soon as the script exits.

**Review by adriansuarez — APPROVED** [review]
Looks good.

**Pull request merged**